bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TypeChecker.java

```diff
@@ -1279,32 +1297,115 @@ private static boolean checkIsArrayType(BType sourceType, BArrayType targetType,
+            sourceTupleType = (BTupleType) sourceType;
+        } else {
+            BArrayType sourceArrayType = (BArrayType) sourceType;
+            BType sourceElementType = sourceArrayType.getElementType();
+
+            switch (sourceArrayType.getState()) {
+                case OPEN:
+                    if (targetRestType == null) {
+                        return false;
+                    }
+                    if (targetTypes.isEmpty()) {
+                        return checkIsType(sourceElementType, targetRestType, unresolvedTypes);
+                    }
+                    return false;
+                case CLOSED:
+                    if (sourceArrayType.getSize() < targetTypes.size()) {
+                        return false;
+                    }
+                    if (targetTypes.isEmpty()) {
+                        if (targetRestType != null) {
+                            return checkIsType(sourceElementType, targetRestType, unresolvedTypes);
+                        }
+                        if (sourceArrayType.getSize() == 0) {
+                            return true;
+                        }
+                        return false;
+                    }
+                    boolean memberTypesMatch = true;
+
+                    for (BType targetElementType : targetTypes) {
+                        if (!(checkIsType(sourceElementType, targetElementType, unresolvedTypes))) {
+                            memberTypesMatch = false;
+                        }
+                    }
+                    if (sourceArrayType.getSize() == targetTypes.size()) {
+                        return memberTypesMatch;
+                    }
+                    if (targetRestType != null) {
+                        return (memberTypesMatch &&
+                                checkIsType(sourceElementType, targetRestType, unresolvedTypes));
+                    }
+                    return false;
+            }
         }
 
-        if (sourceTypes.size() != targetTypes.size()) {
+        List<BType> sourceTypes = new ArrayList<>(sourceTupleType.getTupleTypes());
+        BType sourceRestType = sourceTupleType.getRestType();
+
+        if (sourceRestType != null && targetRestType == null) {
             return false;
         }
+        int sourceTypeSize = sourceTypes.size();
+        int targetTypeSize = targetTypes.size();
 
-        for (int i = 0; i < sourceTypes.size(); i++) {
-            if (!checkIsType(sourceTypes.get(i), targetTypes.get(i), unresolvedTypes)) {
-                return false;
+        if (sourceRestType == null && targetRestType == null && sourceTypeSize != targetTypeSize) {
+            return false;
+        }
+
+        boolean memberTypesMatch = true;
+
+        if (sourceTypeSize >= targetTypeSize) {
+            for (int i = 0; i < targetTypeSize; i++) {
+                if (!checkIsType(sourceTypes.get(i), targetTypes.get(i), unresolvedTypes)) {
+                    memberTypesMatch = false;
```

Comment:
Can't we directly return false here instead of setting it to `memberTypesMatch`?

Reply:
Check `remainingMemberTypesMatch` too.